bundle/config/mutator/python/override.go

```diff
@@ -0,0 +1,313 @@
+package python
+
+import (
+	"fmt"
+
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/dyn/merge"
+)
+
+// overrideMode restricts which changes are allowed during the override process
+type overrideMode string
+
+const OverrideModeAppend overrideMode = "append"
+const OverrideModeAppendOrModify overrideMode = "append-or-update"
+
+// isModified keeps track of whether a value has been modified during the override process
+type isModified struct {
+	value bool
+}
+
+var Modified = isModified{value: true}
+var NotModified = isModified{value: false}
+
+type overrideFunc func(dyn.Value, dyn.Value) (dyn.Value, isModified, error)
+
+// OverrideByPath overrides bundle config 'leftRoot' with 'rightRoot' values in 'path', keeping
+// 'location' if values haven't changed. This allows to keep track of where the values
+// were last modified for error reporting.
+//
+// 'path' must point to a mapping or not exist, otherwise, error is returned
+//
+// 'mode' allows to restrict what kinds of changes are allowed in 'path':
+// - append: only new values can be added to mapping
+// - append-or-modify: new values can be added, or existing values can be modified
+func OverrideByPath(leftRoot dyn.Value, rightRoot dyn.Value, path dyn.Path, mode overrideMode) (dyn.Value, error) {
+	_, leftErr := dyn.GetByPath(leftRoot, path)
+	_, rightErr := dyn.GetByPath(rightRoot, path)
+
+	if leftErr != nil && rightErr != nil {
+		return leftRoot, nil
+	}
+
+	leftRoot, err := insertMappingIfAbsent(leftRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	rightRoot, err = insertMappingIfAbsent(rightRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
```

Comment:
Can't we operate directly on the root values?

It seems like this indirection combines what could otherwise be done by a caller with an additional `dyn.Map` call.

Reply:
I am not sure I got you, but we have to override only specific parts of the root config. For instance, we want to allow adding new variables only during pre-init (not yet implemented), we never want to change "environments" (because we aren't going to pick up changes correctly), and for now, we are only expected to modify "jobs" by appending or changing them. So we really need to handle different parts of bundle config differently.

Reply:
Yeah, I understand the intent.

What I'm proposing it to let the caller deal with this, like this (roughly):
```go
left := dyn.GetByPath(v1, "resources.jobs")
right := dyn.GetByPath(v2, "resources.jobs")
out, status := Override(left, right)
dyn.SetByPath(vout, "resources.jobs", out)
```

Then all logic where we decide what to allow for which paths can be dealt with there and we have a "pure" override function for any `dyn.Value` that returns the right amount of information we need to decide whether it is allowed or not.